drivers/soft_spi/soft_spi.c

```diff
@@ -167,46 +163,112 @@ static inline uint8_t _transfer_one_byte(soft_spi_t bus, uint8_t out)
     return out;
 }
 
-uint8_t soft_spi_transfer_byte(soft_spi_t bus, soft_spi_cs_t cs, bool cont, uint8_t out)
+static uint8_t _readin_one_byte(soft_spi_t bus)
+{
+    int8_t i = 0, tmp = 0;
+    int bit;
+
+    if (SOFT_SPI_MODE_1 == soft_spi_config[bus].soft_spi_mode ||
+        SOFT_SPI_MODE_3 == soft_spi_config[bus].soft_spi_mode) {
+        /* CPHA = 1*/
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+        xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
+    }
+    gpio_toggle(soft_spi_config[bus].clk_pin);
+    xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
+    bit = gpio_read(soft_spi_config[bus].miso_pin);
+    tmp |= (!!bit << 7);
+    for (i = 6; i >= 0; i--) {
+        xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+        xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+        bit = gpio_read(soft_spi_config[bus].miso_pin);
+        tmp |= (!!bit << i);
+    }
+
+    if (SOFT_SPI_MODE_0 == soft_spi_config[bus].soft_spi_mode ||
+        SOFT_SPI_MODE_2 == soft_spi_config[bus].soft_spi_mode) {
+        xtimer_nanosleep(soft_spi_config[bus].soft_spi_clk);
+        gpio_toggle(soft_spi_config[bus].clk_pin);
+    }
+
+    return tmp;
+}
+
+uint8_t soft_spi_transfer_byte(soft_spi_t bus, soft_spi_cs_t cs, bool cont,
+                               uint8_t out)
 {
     DEBUG("Soft SPI soft_spi_transfer_bytes\n");
-    assert(soft_spi_bus_is_valid(bus));
+    assert(bus < SOFT_SPI_NUMOF);
 
     uint8_t retval = 0;
 
     /* activate the given chip select line */
-    if (gpio_is_valid(cs) && !gpio_is_equal(cs, SOFT_SPI_CS_UNDEF)) {
+    if ((cs != GPIO_UNDEF) && (cs != SOFT_SPI_CS_UNDEF)) {
         gpio_clear((gpio_t)cs);
     }
 
     retval = _transfer_one_byte(bus, out);
```

Comment:
`_transfer_one_byte()` needs to be fixed - it currently will only output the byte on `mosi` but not read `miso`.